### PR TITLE
Deprecate read_feather nthreads argument + update feather-format to pyarrow.feather

### LINKED ISSUE #23053: pandas/io/feather_format.py should call use_threads instead of nthreads to prevent breakage in pyarrow 0.11.0
#### Code Sample

```python
d = {'one' : [1., 2., 3., 4.],
        'two' : [4., 3., 2., 1.]}
df = pandas.DataFrame(d)
df.to_feather('example.feather')

# with pyarrow 0.10.0 this succeeds with a deprecation warning
# with pyarrow 0.11.0 this errors with a TypeError: unexpected argument 'nthreads'
df = pandas.read_feather('example.feather')

# attempt to manually set nthreads results in TypeError: unexpectect argument 'nthreads'
df = pandas.read_feather('example.feather', nthreads=4)

# attempt to pass 'use_threads' results in TypeError: unexpected argument 'nthreads'
df = pandas.read_feather('example.feather', use_threads=True)

```
#### Problem description
Pandas introduced nthreads for reading feather files in issue [16359](https://github.com/pandas-dev/pandas/issues/16359)

With PyArrow 0.10.0 a deprecation warning is shown from [this source](https://github.com/apache/arrow/blob/master/python/pyarrow/util.py): "`nthreads` argument is deprecated, pass `use_threads` instead"

When PyArrow version 0.11.0, Python errors with: TypeError: read_feather() got an unexpected keyword argument 'nthreads'.

I've searched with 'pyarrow' and 'nthreads' keywords and didn't see this issue posted.

Specifically [feather-format.py line 112](https://github.com/pandas-dev/pandas/blob/master/pandas/io/feather_format.py#L112) should be changed to 
`return feather.read_dataframe(path, use_threads=True)` or changing the method signature to all overriding use_threads:
`return feather.read_dataframe(path, use_threads=use_threads)`
I will submit a PR if the only barrier to fix is code effort.
#### Expected Output
I expect no error output upon running pandas.read_feather() with PyArrow 0.11.0

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS                                           
------------------                                           
commit: None                                                 
python: 3.6.6.final.0                                        
python-bits: 64                                              
OS: Windows                                                  
OS-release: 7                                                
machine: AMD64                                               
processor: Intel64 Family 6 Model 79 Stepping 1, GenuineIntel
byteorder: little                                            
LC_ALL: None                                                 
LANG: None                                                   
LOCALE: None.None                                            
                                                             
pandas: 0.23.4                                               
pytest: None                                                 
pip: 18.1                                                    
setuptools: 40.3.0                                           
Cython: None                                                 
numpy: 1.15.1                                                
scipy: 1.1.0                                                 
pyarrow: 0.10.0                                              
xarray: None                                                 
IPython: 6.5.0                                               
sphinx: None                                                 
patsy: 0.5.0                                                 
dateutil: 2.7.3                                              
pytz: 2018.5                                                 
blosc: None                                                  
bottleneck: None                                             
tables: None                                                 
numexpr: None                                                
feather: 0.4.0                                               
matplotlib: 2.2.2                                            
openpyxl: None                                               
xlrd: 1.1.0                                                  
xlwt: None                                                   
xlsxwriter: None                                             
lxml: None                                                   
bs4: None                                                    
html5lib: None                                               
sqlalchemy: 1.2.11                                           
pymysql: None                                                
psycopg2: None                                               
jinja2: None                                                 
s3fs: None                                                   
fastparquet: None                                            
pandas_gbq: None                                             
pandas_datareader: None                                      

</details>

### PR DESCRIPTION
**Pull request by ingwinlu**
- [x] closes #23053, closes https://github.com/pandas-dev/pandas/issues/21639
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
this would need a whatsnew note
and i believe this is tested so pls update the tests and catch the warning

**Comment by jorisvandenbossche** [comment]
Shouldn't it be replaced with `use_threads` ?

**Comment by ingwinlu** [comment]
this would require pinning pyarrow > 0.10.0 as a dependency

File: doc/source/whatsnew/v0.23.5.txt [hunk 1]
```diff
@@ -33,6 +33,8 @@ Fixed Regressions
 Development
 ~~~~~~~~~~~
 - The minimum required pytest version has been increased to 3.6 (:issue:`22319`)
+- Deprecated the `nthreads` keyword of `pandas.read_feather()` in favor of
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.5.txt]: use the :func: reference here

File: pandas/io/feather_format.py [hunk 2]
```diff
@@ -96,6 +101,11 @@ def read_feather(path, nthreads=1):
         Number of CPU threads to use when reading to pandas.DataFrame
 
        .. versionadded 0.21.0
+       .. deprecated 0.23.5
```
Comment (jreback) [inline pandas/io/feather_format.py]: 0.24.0

File: pandas/io/feather_format.py [hunk 3]
```diff
@@ -106,7 +116,4 @@ def read_feather(path, nthreads=1):
     feather = _try_import()
     path = _stringify_path(path)
 
-    if LooseVersion(feather.__version__) < LooseVersion('0.4.0'):
```
Comment (jreback) [inline pandas/io/feather_format.py]: why is this removed?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/feather_format.py (hunk 3 above)
Comment (ingwinlu) [reply pandas/io/feather_format.py]: I wasn't sure if we would still need it if we would pin a higher version of pyarrow.

File: pandas/io/feather_format.py [hunk 4]
```diff
@@ -96,17 +103,23 @@ def read_feather(path, nthreads=1):
...
     type of object stored in file
 
     """
 
-    feather = _try_import()
+    feather, pyarrow = _try_import()
     path = _stringify_path(path)
 
-    if LooseVersion(feather.__version__) < LooseVersion('0.4.0'):
+    if LooseVersion(feather.__version__) < LooseVersion('0.4.0') or \
```
Comment (TomAugspurger) [inline pandas/io/feather_format.py]: Style nit: start the line with parentheses instead of using a backslash to continue the line.

File: doc/source/whatsnew/v0.24.0.txt [hunk 5]
```diff
@@ -979,3 +979,5 @@ Other
 - :meth:`~pandas.io.formats.style.Styler.bar` now also supports tablewise application (in addition to rowwise and columnwise) with ``axis=None`` and setting clipping range with ``vmin`` and ``vmax`` (:issue:`21548` and :issue:`21526`). ``NaN`` values are also handled properly.
 - Logical operations ``&, |, ^`` between :class:`Series` and :class:`Index` will no longer raise ``ValueError`` (:issue:`22092`)
 - Bug in :meth:`DataFrame.combine_first` in which column types were unexpectedly converted to float (:issue:`20699`)
+- Deprecated the `nthreads` keyword of :func:`pandas.read_feather()` in favor of
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.0.txt]: Don't need the parentheses in `read_feather()`

File: pandas/io/feather_format.py (hunk 4 above)
Comment (ingwinlu) [reply pandas/io/feather_format.py]: Do you have a suggestion for the alternative? It does not line up as nicely (since it would require an additional indent of the second line). Making it in my opinion harder to read.

Comment (TomAugspurger) [reply pandas/io/feather_format.py]:
```
if (LooseVersion(feather.__version__) < LooseVersion('0.4.0') or
        LooseVersion(pyarrow.__version__) < LooseVersion('0.11.0')):
    return feather.read_dataframe(path)
```

**Review by jreback — CHANGES_REQUESTED** [review]
does this emit any warnings in the tests?

I would actually be ok with removing support for feather < 0.4.0 if it makes it easier here. you may actually *need* to do this as importing an older feather doesn't have pyarrow as a dep so the importer might fail.

**Comment by jreback** [comment]
@ingwinlu actually I will revise the though above. Am ok with dropping support for feather entirely and just using pyarrow here. Can you revise?

**Comment by ingwinlu** [comment]
Sure. Will probably work on it later today.

**Comment by ingwinlu** [comment]
@jreback do you want to depend on pyarrow 0.4.0 or go for 0.11.0 directly?

**Comment by jreback** [comment]
if u can use our min version of pyarrow (we use for parquet)

we could bump that slightly also but. o past 0.8.0

**Comment by ingwinlu** [comment]
@jreback done. don't think the windows test fail is related to the changes in the PR.

**Comment by ingwinlu** [comment]
This is still missing the removal of feather-format refs in ci configs as well as some doc entry in source/io I missed.

**Comment by h-vetinari** [comment]
If dropping feather, will also close #21639

**Comment by ingwinlu** [comment]
I did not rewrite the io.feather section of the docs. I feel like the currently linked repository (feather) provides more information.

If you feel like it is necessary we can add a point to the caveats listed where we express that we directly require the upstream pyarrow library.

File: doc/source/whatsnew/v0.24.0.txt [hunk 6]
```diff
@@ -1004,3 +1004,7 @@ Other
 - :meth:`~pandas.io.formats.style.Styler.bar` now also supports tablewise application (in addition to rowwise and columnwise) with ``axis=None`` and setting clipping range with ``vmin`` and ``vmax`` (:issue:`21548` and :issue:`21526`). ``NaN`` values are also handled properly.
 - Logical operations ``&, |, ^`` between :class:`Series` and :class:`Index` will no longer raise ``ValueError`` (:issue:`22092`)
 - Bug in :meth:`DataFrame.combine_first` in which column types were unexpectedly converted to float (:issue:`20699`)
+- Deprecated the `nthreads` keyword of :func:`pandas.read_feather` in favor of
+  `use_threads` to reflect the changes in pyarrow 0.11.0. (:issue:`23053`)
+- Drop `feather-format` as a dependency for feather based storage and use
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: move this to api-breaking changes section

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.0.txt [hunk 7]
```diff
@@ -233,6 +233,10 @@ If installed, we now require:
 | scipy           | 0.18.1          |          |
 +-----------------+-----------------+----------+
 
+
+Additionally we no longer depend on `feather-format` for feather based storage
+and replaced it with references to `pyarrow`.
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: add the issue refernce here

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
going to merge after passing: https://github.com/pandas-dev/pandas/pull/23281, which you will need to revert the xfails.

**Comment by jreback** [comment]
can you rebase this and check changes in #23281

File: doc/source/whatsnew/v0.24.0.txt [hunk 8]
```diff
@@ -236,6 +236,10 @@ If installed, we now require:
 | scipy           | 0.18.1          |          |
 +-----------------+-----------------+----------+
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: you can add pyarrow to the table above with a min of 0.4.1 (you comment is ok to leave)

File: ci/travis-36.yaml [hunk 9]
```diff
@@ -7,7 +7,6 @@ dependencies:
   - cython>=0.28.2
   - dask
   - fastparquet
-  - feather-format
```
Comment (jreback) [inline ci/travis-36.yaml]: can you fix pyarrow=0.10.0 here, and add pyarrow to the 3.7 build

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. let's just add a prior to change pyarrow version to test (0.10.0), ping on green.

File: doc/source/whatsnew/v0.24.0.txt [hunk 10]
```diff
@@ -235,6 +235,11 @@ If installed, we now require:
...
 | scipy           | 0.18.1          |          |
 +-----------------+-----------------+----------+
+| pyarrow         | 0.4.1           |          |
```
[the remark this answers is not in the document]
Comment (jreback) [reply doc/source/whatsnew/v0.24.0.txt]: yes, but not if you don't have feather

**Comment by jreback** [comment]
can you rebase and fixup?

**Comment by ingwinlu** [comment]
I can not replicate the 
```
>   import pyarrow.compat as compat
E   AttributeError: module 'pyarrow' has no attribute 'compat'
```
issues locally. Any pointers?

**Comment by TomAugspurger** [comment]
Maybe a different version of pyarrow? https://travis-ci.org/pandas-dev/pandas/jobs/446626253#L2219

**Comment by ingwinlu** [comment]
I replicated the conda environment of that test run. parquet tests run correctly but feather tests are not run at all due to in import arrow on pyarrow:
```
Python 3.6.6 |Anaconda, Inc.| (default, Oct  9 2018, 12:34:16) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyarrow
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/winlu/miniconda3/envs/pandas/lib/python3.6/site-packages/pyarrow/__init__.py", line 60, in <module>
    from pyarrow.lib import cpu_count, set_cpu_count
ImportError: /home/winlu/miniconda3/envs/pandas/lib/python3.6/site-packages/pyarrow/../../../libparquet.so.1: undefined symbol: _ZN5boost13match_resultsIN9__gnu_cxx17__normal_iteratorIPKcSsEESaINS_9sub_matchIS5_EEEE12maybe_assignERKS9_
>>> 
```

Could that be some caching issue?

**Comment by ingwinlu** [comment]
@jreback should be good now.

File: pandas/io/feather_format.py [hunk 11]
```diff
@@ -96,17 +97,22 @@ def read_feather(path, nthreads=1):
...
     """
 
-    feather = _try_import()
+    feather, pyarrow = _try_import()
     path = _stringify_path(path)
 
-    if LooseVersion(feather.__version__) < LooseVersion('0.4.0'):
-        return feather.read_dataframe(path)
+    if LooseVersion(pyarrow.__version__) < LooseVersion('0.11.0'):
+        return feather.read_feather(path)
```
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply pandas/io/feather_format.py]: @ingwinlu can you check this comment?

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]

File: pandas/io/feather_format.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ingwinlu) [reply pandas/io/feather_format.py]: nthreads is not available anymore after the conversion in the wrapper function.

[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply pandas/io/feather_format.py]: Then we need to do it in some other way, but not passing it here means breaking the functionality for people having pyarrow < 0.11 (which we still support)

File: doc/source/whatsnew/v0.24.0.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (jreback) [reply doc/source/whatsnew/v0.24.0.txt]: @ingwinlu actually maybe take out pyarrow from the table, your comment below is fine

File: pandas/io/feather_format.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/io/feather_format.py]: @ingwinlu what you could do is remove the ``mapping`` part of the deprecation, then you can easily just pass the argument here as ``read_feather(path, nthreads=int(use_threads))`` I think would work, then on the other branch, just pass ``read_feather(path, use_threads=bool(use_threads))``

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@ingwinlu thanks for the responsiveness. @jorisvandenbossche over to you.

**Comment by jreback** [comment]
can you rebase once more.

**Comment by ingwinlu** [comment]
done

File: pandas/io/feather_format.py [hunk 12]
```diff
@@ -99,17 +97,25 @@ def read_feather(path, nthreads=1):
...
-    if LooseVersion(feather.__version__) < LooseVersion('0.4.0'):
-        return feather.read_dataframe(path)
+    if LooseVersion(pyarrow.__version__) < LooseVersion('0.11.0'):
+        int_use_threads = int(use_threads)
+        if int_use_threads < 1:
+            int_use_threads = 1
+        return feather.read_feather(path, nthreads=int_use_threads)
 
-    return feather.read_dataframe(path, nthreads=nthreads)
+    return feather.read_feather(path, use_threads=bool(use_threads))
```
Comment (jorisvandenbossche) [inline pandas/io/feather_format.py]: I don't think this is fully correct. If someone did before `nthreads=1` (which meant: no additional threads), this will be translated into `use_threads=True`.

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]

File: pandas/io/feather_format.py (hunk 12 above)
Comment (jorisvandenbossche) [reply pandas/io/feather_format.py]: Although, maybe that is not really a problem since the default in pyarrow also actually changed from `nthreads=1` to `use_threads=True`

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
@ingwinlu Thanks a lot!
